include/boost/gil/io/detail/filesystem.hpp

```diff
@@ -0,0 +1,65 @@
+//
+// Copyright 2022 Mateusz Loskot <mateusz at loskot dot net>
+//
+// Distributed under the Boost Software License, Version 1.0
+// See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt
+//
+#ifndef BOOST_GIL_IO_DETAIL_FILESYSTEM_HPP
+#define BOOST_GIL_IO_DETAIL_FILESYSTEM_HPP
+
+#include <boost/config.hpp>
+
+#if !defined(BOOST_GIL_IO_USE_BOOST_FILESYSTEM)
+#if !defined(BOOST_NO_CXX17_HDR_FILESYSTEM) || defined(__cpp_lib_filesystem)
```

Comment:
If I understood correctly, it seems like `BOOST_NO_CXX17_HDR_FILESYSTEM` is the boost wide flag and the one above is GIL specific flag, right? Perhaps we should somehow enforce consistency between the two? Also, it seems like the flags `BOOST_GIL_IO_USE_STD_FILESYSTEM`, `BOOST_GIL_IO_USE_STD_EXPERIMENTAL_FILESYSTEM` are not orthogonal. Perhaps we could have one flag that has 3 distinct values (BOOST_FILESYSTEM, STD_FILESYSTEM, STD_EXPERIMENTAL_FILESYSTEM) and ifdef on those?

Reply:
> it seems like BOOST_NO_CXX17_HDR_FILESYSTEM is the boost wide flag and the one above is GIL specific flag, right?

Yes

> Perhaps we should somehow enforce consistency between the two?

Not sure what you mean.

>  it seems like the flags BOOST_GIL_IO_USE_STD_FILESYSTEM, BOOST_GIL_IO_USE_STD_EXPERIMENTAL_FILESYSTEM are not orthogonal.

Rather, the latter is a part (subset) of the former.

> Perhaps we could have one flag that has 3 distinct values

If it wasn't a preprocessor but e.g. Python, I'd love that idea :)
With preprocessor, I tend to keep things dead simple, distinct names, which can be added or removed in future - IMO, it's easier to validate if a preprocessor macro exists, than if a set of values of a macro changed.